sklearn/preprocessing/imputation.py

```diff
@@ -375,3 +375,203 @@ def transform(self, X):
             X[coordinates] = values
 
         return X
+
+
+class MissingIndicator(BaseEstimator, TransformerMixin):
+    """Missing values indicator transformer for indicating missing values.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default="NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    features : {'train' (default), 'all', array-like of int}
+        If "all", mask will represent all features.
+        If "train", mask will only represent features with missing values
+        during fit time.
+        If mask/indices, mask will only represent features in the
+        indices or mask.
+
+    sparse : boolean or "auto", optional (default="auto")
+        If True, the transformed ``X`` will be a sparse matrix.
+        If False, the transformed ``X`` will be a numpy array.
+        If "auto", the transformed ``X`` will be of same type as input.
+
+    Example
+    -------
+    >>> from sklearn.preprocessing import MissingIndicator
+    >>> import numpy as np
+    >>> X1 = np.array([
+    ...       [-1,  1,  3],
+    ...       [ 4,  0, -1],
+    ...       [ 8,  1,  0]
+    ... ])
+    >>> X2 = np.array([
+    ...       [ 5, -1, -1],
+    ...       [-1,  2,  3],
+    ...       [ 2,  4,  0]
+    ... ])
+    >>> indicator = MissingIndicator(missing_values=-1)
+    >>> indicator.fit(X1)
+    MissingIndicator(features='train', missing_values=-1, sparse='auto')
+    >>> X2_tr = indicator.transform(X2)
+    >>> X2_tr
+    array([[0, 1],
+           [1, 0],
+           [0, 0]])
+
+    Attributes
+    ----------
+    feat_with_missing_  : array of shape(n_missing_features, )
+        The features with missing values.
+        Note that this is only stored if features == 'train
+
+    n_features_ : int
+        The number of features during fit time.
+    """
+
+    def __init__(self, missing_values="NaN", features="train", sparse="auto"):
+        self.missing_values = missing_values
+        self.features = features
+        self.sparse = sparse
+
+    def fit(self, X, y=None):
+        """Fit the transformer on X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        self._validate_params()
+
+        X = check_array(X, accept_sparse=('csc', 'csr'), dtype=np.float64)
+        self.n_features_ = X.shape[1]
+
+        if (isinstance(self.features, six.string_types) and
+                self.features == "train"):
+            _, self.feat_with_missing_ = self._get_missing_features_info(X)
```

Comment:
And just a thought, even if this makes the code simpler, we will need to handle the case when `self.features = "all"` before slicing the mask. Correct me please if I am missing something.